client: split CommitTimestamp() from OrigTimestamp()

vivekmenezes opened the pull request:
Also get rid of OrigTimestampWasObserved()

Release note: None

vivekmenezes committed:
client: split CommitTimestamp() from OrigTimestamp()

Also get rid of OrigTimestampWasObserved()

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/22932)
<!-- Reviewable:end -->

knz commented:
Ok for the change to `OrigTimestamp` and introducing `CommitTimestamp`. I think we can settle on the idea that any caller of `CommitTimestamp` also cares about that value staying stable.

However I do believe that the method `OrigTimestampWasObserved` must remain, perhaps renamed to `FixCommitTimestamp()` (no return value). Otherwise the code reads ugly, see below.

---

Reviewed 12 of 12 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/kv/dist_sender_server_test.go, line 2454 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22932#-L5y-xs1Bp1iRFpxmV8y:-L5y-xs2dyopTPFnBMet:b-cds5x7) ([raw file](https://github.com/cockroachdb/cockroach/blob/74ca3e7d993c4745f699c16d60151c614d946749/pkg/kv/dist_sender_server_test.go#L2454)):*
> ```Go
> 			if err := db.Txn(ctx, func(ctx context.Context, txn *client.Txn) error {
> 				if tc.tsLeaked {
> 					_ = txn.CommitTimestamp()
> ```

To be honest I really don't like this.
The previous form made it clear there was a side effect (all functions that return `void` can only exist because they do side effects). This new form doesn't.

I'd be ok reading instead `txn.FixCommitTimestamp()`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22932#-:-L5y0PGaQdUtBN5R6uY8:b-qn0nfk)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
TFTR! See my comment below

---

Review status: 11 of 12 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/kv/dist_sender_server_test.go, line 2454 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22932#-L5y-xs1Bp1iRFpxmV8y:-L5y1d8wloCR2JQn50ow:b-vn97ip) ([raw file](https://github.com/cockroachdb/cockroach/blob/74ca3e7d993c4745f699c16d60151c614d946749/pkg/kv/dist_sender_server_test.go#L2454)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

To be honest I really don't like this.
The previous form made it clear there was a side effect (all functions that return `void` can only exist because they do side effects). This new form doesn't.

I'd be ok reading instead `txn.FixCommitTimestamp()`.
</blockquote></details>

The reason why I removed it is because that method was originally used but is now only used in this test. I hate adding a method to an API for a test unless it's absolutely needed for testing.  I've added a comment explaining the use of CommitTimestamp()

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22932#-:-L5y2nckzaBI3OHa2HsK:b-boyof7)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

Thank you!!

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/kv/dist_sender_server_test.go, line 2454 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22932#-L5y-xs1Bp1iRFpxmV8y:-L5y3SGKgal1mnDVnsLu:b-9zt01x) ([raw file](https://github.com/cockroachdb/cockroach/blob/74ca3e7d993c4745f699c16d60151c614d946749/pkg/kv/dist_sender_server_test.go#L2454)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

The reason why I removed it is because that method was originally used but is now only used in this test. I hate adding a method to an API for a test unless it's absolutely needed for testing.  I've added a comment explaining the use of CommitTimestamp()
</blockquote></details>

Ok I can live with that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22932#-:-L5y3UU-m1zbipf064g9:b-dr7vb8)*
<!-- Sent from Reviewable.io -->
